src/SonarScanner.MSBuild.PreProcessor/JreCaching/JreCache.cs

```diff
@@ -139,15 +159,59 @@ private static void EnsureClosed(Stream fileStream)
         }
     }
 
-    private async Task<JreCacheResult> UnpackJre(string downloadTarget, JreDescriptor jreDescriptor, string cacheRootLocation)
+    private JreCacheResult ValidateAndUnpackJre(IUnpack unpacker, string jreArchive, JreDescriptor jreDescriptor, string cacheRoot)
     {
-        if (ValidateChecksum(downloadTarget, jreDescriptor.Sha256))
+        if (ValidateChecksum(jreArchive, jreDescriptor.Sha256))
         {
-            return new JreCacheFailure("NotImplemented. The JRE is downloaded and validated, but we still need to unpack, and set permissions.");
+            return UnpackJre(unpacker, jreArchive, jreDescriptor, cacheRoot);
         }
```

Comment:
I'm not sure why you extracted this validation outside. `UnpackJre` is only used by this method. I prefer the old version, where the validation is implicit and it does not surface in the name.